wolfssl/test.h

```diff
@@ -3873,6 +3873,49 @@ static WC_INLINE int myEccSharedSecret(WOLFSSL* ssl, ecc_key* otherKey,
 
 #endif /* HAVE_ECC */
 
+#ifdef HAVE_HKDF
+static WC_INLINE int myHkdfExtract(byte* prk, const byte* salt, word32 saltLen,
+       byte* ikm, word32 ikmLen, int digest, void* ctx)
+{
+    int ret;
+    int len = 0;
+
+    switch (digest) {
+#ifndef NO_SHA256
+        case WC_SHA256:
+            len = WC_SHA256_DIGEST_SIZE;
+            break;
+#endif
+
+#ifdef WOLFSSL_SHA384
+        case WC_SHA384:
+            len = WC_SHA384_DIGEST_SIZE;
+            break;
+#endif
+
+#ifdef WOLFSSL_TLS13_SHA512
+        case WC_SHA512:
+            len = WC_SHA512_DIGEST_SIZE;
+            break;
+#endif
+        default:
+            return BAD_FUNC_ARG;
+    }
+
+    /* When length is 0 then use zeroed data of digest length. */
+    if (ikmLen == 0) {
+        ikmLen = len;
+        XMEMSET(ikm, 0, len);
+    }
+
+    (void)ctx;
+    ret = wc_HKDF_Extract(digest, salt, saltLen, ikm, ikmLen, prk);
+    WOLFSSL_PKMSG("PK HKDF Extract: ret %d saltLen %d ikmLen %d\n", ret, saltLen,
+            ikmLen);
+    return ret;
+}
```

Comment:
Since this is a copy of `wc_Tls13_HKDF_Extract`, I think it would be better to just call that instead.

Reply:
There is a reason for the duplication in `test.h`:
Other PK_CALLBACK tests in the same file re-implement the top level of the function by copying the actual code. This ensures that we can scope the callbacks' steps (e.g. when running tests with `-P`) that do not just call the equivalent function being replaced by the callback itself.

Reply:
Can you expand on what you mean with "scope the callbacks' steps"? A breakpoint could still be set in this test function but the implementation, instead of copying `wc_Tls13_HKDF_Extract` would just call it directly (maybe after printing some logs about a callback success).

Reply:
It's equivalent from the code standpoint. I just did not want to override the call with a callback that call itself. Let me know if it's an issue and I'll replace the body of the test.h callback with a single call to it anyway.

Reply:
I still think that the test.h callback should show what operations are done with the data before calling the actual HKDF_Extract function from wolfcrypt, so it would be clearer to have the code here for reference, for people implementing a similar functionality (this way it's clear that the operation underneath may require, like the wolfcrypt one, to pass a zeroed buffer as ikm, where the length depends on the digest).

I'm open to both solutions tho, but test.h is verbose on purpose for other replacement callbacks, that can be used as base implementation for overriding these functions using a different engine.

Reply:
I'll let the final reviewer make the call. Leaving this conversation open.